design_notes/XDN03 - xlang Type System.md

```diff
@@ -15,3 +15,142 @@ status: draft
 This design note describes the xlang's [type system](https://en.wikipedia.org/wiki/Type_system).
 The type system is designed to interoperate with a variety of object-oriented programming
 languages in a natural and familiar way.
+
+## Namespaces
+
+A namespace is a naming scope used to organize code and avoid naming collisions. All type
+categories in the xlang type system (enums, structs, delegates, interfaces, and runtime classes)
+except for fundamental types live in a namespace.
+
+Namespaces may contain other namespaces.
+
+xlang namespaces and type names are case preserving but insensitive. This means that you cannot
+have namespaces or type names that vary only by case. For example: you cannot have `Foo.SomeType`
+and `foo.SomeType` nor can you have `Foo.SomeType` and `Foo.someType`. However, you can have
+`Foo.SomeType` and `foo.AnotherType`. In this case `SomeType` and `AnotherType` would be considered
+as belonging to the same namespace.
+
+## Fundamental Types
+
+The xlang Type system includes a core set of built-in primitive types.
+
+xlang Type | Type Description
+---------- | ----------------------------------------------------------
+Int16      | a 16 bit signed integer
+Int32      | a 32 bit signed integer
+Int64      | a 64 bit signed integer
+UInt8      | an 8 bit unsigned integer
+UInt16     | a 16 bit unsigned integer
+UInt32     | a 32 bit unsigned integer
+UInt64     | a 64 bit unsigned integer
+Single     | a 32-bit IEEE 754 floating point number
+Double     | a 64-bit IEEE 754 floating point number
+Char16     | a 16-bit non-numeric value representing a UTF-16 code unit
+Boolean    | an 8-bit Boolean value
+String     | an immutable sequence of Char16s used to represent text
+Guid       | A 128-bit standard [universally unique identifier](https://en.wikipedia.org/wiki/Universally_unique_identifier)
+Object     | An xlang object of unknown type
+
+> Note: string encoding for xlang is currently under review. In particular, it has not been decided
+if xlang will support UTF-8 strings in some capacity. A design note will be commissioned for this
```

Comment:
It would be nice to _only_ support UTF-8 and do silent conversions for back-compat. I think I heard .NET Core is getting a UTF-8 string type, so it'd be real nice to have a way to keep it consistent.

Reply:
Yeah, I'd like to only support UTF-8 as well. Design note for string encoding will dig into this further. 

I seem to remember something about a UTF-8 string type for .NET, but not sure now where I saw it.

Reply:
https://github.com/dotnet/corefxlab/issues/2350

Reply:
We will need both.  UTF16 is a requirement for languages like Java, which use it natively.  Otherwise, we unnecessarily convert back and forth.

Reply:
@DHowett-MSFT might be able to provide some insight on how this sort of thing was handled in Objective-C. I believe it *mostly* hides the internal representation from the developer except for exposing a very few "raw" functions that operate on the individual characters etc. This would, however, mean building a more full featured string class as part of the "foundation classes"

Reply:
Looks like JNI supports both 8 and 16 bit strings as well. [NewString](https://docs.oracle.com/javase/7/docs/technotes/guides/jni/spec/functions.html#wp4925) constructs a jstring from an array of 16-bit jchars while [NewStringUTF](https://docs.oracle.com/javase/7/docs/technotes/guides/jni/spec/functions.html#wp17220) constructs a jstring from an array of 8-bit chars.

Reply:
https://stackoverflow.com/a/9699138/500701

Reply:
Rephrased the note and added a link to issue #53

Reply:
Lets use #53 for further discussion